File: HW7/BoardArray1D.java
import java.io.*;
import java.security.SecureRandom;
import java.util.Scanner;

/**
 * Class BoardArray1D is a subclass of AbstractBoard class
 * Implements all methods of super class
 * Keeps board configuration at 1D array
 */
public class BoardArray1D extends AbstractBoard{

    private int[] board;
    private int row;
    private int column;

    public void setRow(int n_row){
        this.row=n_row;
    }
    public void setColumn(int n_column){
        this.column=n_column;
    }
    public int getRow(){
        return this.row;
    }
    public int getColumn(){
        return this.column;
    }

    /**
     * No parameter constructor
     */
    public BoardArray1D(){
        this.row=9;
        this.column=9;
        board = new int[row*column];
        for(int i=0;i<row*column;i++){
            board[i]=-1;
        }
        setNumOfBoard();
        setLastMove('S');
    }

    /**
     * Constructor that takes size information
     * @param size for square shape board
     */
    public BoardArray1D(int size){
        this.row=size;
        this.column=size;
        board = new int[size*size];
        for(int i=0;i<row*column;i++){
            board[i]=-1;
        }
        setNumOfBoard();
        setLastMove('S');
    }

    /**
     * Constructor that takes two parameters
     * @param n_row as row size
     * @param n_column as column size
     */
    public BoardArray1D(int n_row,int n_column){
        this.row=n_row;
        this.column=n_column;
        board = new int[n_column*n_row];
        for(int i=0;i<row*column;i++){
            board[i]=-1;
        }
        setNumOfBoard();
        setLastMove('S');
    }

    /**
     * toString method to print board configuration
     * @return String board configuration
     */
    @Override
    public String toString(){

        String result="";
        for (int m = 0; m < getRow(); m++){
            for (int k = 0; k < getColumn(); k++) {
                if(board[m*getColumn()+k]!=100)
                    result+= board[m * getColumn() + k] +" ";
                else if(board[m*getColumn()+k]==100)
                    result+= "bb ";
            }
            result+="\n";
        }
        return result;
    }

    /**
     * readFromFile reads file ans assigns the values to board array
     * @param filename as name of the file
     */
    @Override
    public void readFromFile(String filename){
        String[] readen =new String[100];
        try {
            int iRow=0;
            int jCol=0;
            int indexReaden=0;
            Scanner sc = new Scanner(new BufferedReader(new FileReader(filename)));
            //Reading line by line and parsing those lines
            while (sc.hasNextLine()) {
                String[] line = sc.nextLine().split(" ");
                for(int i=0;i<line.length;i++) {
                    jCol=line.length;
                    line[i] = line[i].trim();
                    readen[indexReaden]=line[i];
                    indexReaden++;
                }
               iRow++;
            }
            //Updating board arrays size according to the file
            setSize(iRow,jCol);
            //Assigning readen values to board array
            for(int i=0;i<getRow();i++){
                for(int j=0;j<getColumn();j++){
                    try {
                        //Converting string to integer
                        if (readen[i * getColumn() + j] != "bb")
                            board[i * getColumn() + j] = Integer.parseInt(readen[i * getColumn() + j]);
                    }
                    catch (NumberFormatException nfe){
                        board[i * getColumn() + j] = 100;
                    }
                }
            }
        }
        catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    /**
     * writeToFile method to load board configuration
     * @param filename as name of the file
     */
    @Override
    public void writeToFile(String filename){
        try {
            FileOutputStream output = new FileOutputStream("output.txt");
            output.write(toString().getBytes());
            output.close();
        } catch (IOException e) {
            System.out.println("Error - " + e.toString());
        }
    }

    /**
     * reset method,resets board to the solution
     */
    @Override
    void reset() {
        int num=1;
        for(int i=0;i<getRow();i++){
            for(int j=0; j<getColumn();j++){

                if(i==getRow()-1 && j==getColumn()-1 && board[i*getColumn()+j]!=0){
                    board[i*getColumn()+j]=100;
                }
                else if(board[i*getColumn()+j]!=0){
                    board[i*getColumn()+j]=num;
                    num++;
                }
            }
        }
    }

    /**
     * setSize sets size,and resets the board
     * @param size1 as row
     * @param size2 as column
     */
    @Override
    void setSize(int size1, int size2) {
        setRow(size1);
        setColumn(size2);
        board = new int[size1*size2];
        for(int i=0;i<row*column;i++){
            board[i]=-1;
        }
        reset();
    }

    /**
     * move empty cell if there is a place
     * @param direction U R L D directions
     * @return 1 if move is done,0 if no place
     */
    @Override
    int move(char direction) {
        int emptCellX=-2;
        int emptCellY=-2;
        int num;

        for(int i=0; i<getRow(); i++){
            for(int j=0; j<getColumn(); j++){
                if(board[i*getColumn()+j]==100){
                    emptCellX=i;
                    emptCellY=j;
                }
            }
        }
        if (direction=='U'){

            if(emptCellX==0 || board[((emptCellX-1))*getColumn()+emptCellY]==0){
                return -1;
            }
            else if(emptCellX!=0 && board[((emptCellX-1))*getColumn()+emptCellY]!=0){
                num=board[((emptCellX-1))*getColumn()+emptCellY];
                board[(emptCellX-1)*getColumn()+emptCellY]=100;
                board[emptCellX*getColumn()+emptCellY]=num;
                setLastMove('U');
                setNumOfMove();
            }
            return 0;
        }
        else if(direction== 'D'){

            if(emptCellX==getRow()-1 || board[(emptCellX+1)*getColumn()+emptCellY]==0){
                return -1;
            }
            else if(emptCellX!=getRow()-1 && board[(emptCellX+1)*getColumn()+emptCellY]!=0){
                num=board[(emptCellX+1)*getColumn()+emptCellY];
                board[(emptCellX+1)*getColumn()+emptCellY]=100;
                board[emptCellX*getColumn()+emptCellY]=num;
                setLastMove('D');
                setNumOfMove();
            }
            return 0;
        }
        else if(direction=='R'){

            if(emptCellY==getColumn()-1 || board[emptCellX*getColumn()+(emptCellY+1)]==0){
                return -1;
            }

            else if(emptCellY!=getColumn()-1 && board[emptCellX*getColumn()+(emptCellY+1)]!=0){
                num=board[emptCellX*getColumn()+(emptCellY+1)];
                board[emptCellX*getColumn()+(emptCellY+1)]=100;
                board[emptCellX*getColumn()+emptCellY]=num;
                setLastMove('R');
                setNumOfMove();
            }
            return 0;
        }
        else if(direction=='L'){

            if(emptCellY==0 || board[emptCellX*getColumn()+(emptCellY-1)]==0){
                return -1;
            }

            else if(emptCellY!=0 && board[emptCellX*getColumn()+(emptCellY-1)]!=0){
                num=board[emptCellX*getColumn()+(emptCellY-1)];
                board[emptCellX*getColumn()+(emptCellY-1)]=100;
                board[emptCellX*getColumn()+emptCellY]=num;
                setLastMove('L');
                setNumOfMove();
            }
            return 0;
        }
        return 0;
    }

    /**
     * shuffle method,written to shuffle the board
     * @return int 0 if shuffle is done
     */
    public int shuffle(){

        /*First reset the board to the solution*/
        reset();
        //Initialize the random object
        SecureRandom random = new SecureRandom();

        int randomMove;
        int control=0;
        int i=0;
		/*This loop written for shuffle the board making ramdom moves size*size.
		But if there is an impossible move, i call random func. again to make sure
		the board is shuffled properly */
        while( i<getRow()*getColumn()){

            randomMove = random.nextInt(4) + 1;
            if(randomMove==1){
                control=move('U');
            }
            if(randomMove==2){
                control=move('D');
            }
            if(randomMove==3){
                control=move('L');
            }
            if(randomMove==4){
                control=move('R');
            }
		/*Control is the return value for U,D,R,L functions
		If control is -1 this means there is no room for that choice
		If control is 0 this means cell is moved and size*size number increases*/
            if(control==0)
                i++;
        }
        return 0;
    }

    /**
     * isSolved method to check if board is solution
     * @return true if board is solution,else false
     */
    @Override
    boolean isSolved() {
        BoardArray1D controlObje= new BoardArray1D (row,column);
        controlObje.reset();

        int count=0;
        for(int i=0; i<row; i++){
            for(int j=0; j<column; j++){
                if(board[i*column+j]==controlObje.cell(i,j)){
                    count++;
                }
                else if(board[i*column+j]!=controlObje.cell(i,j)){
                    System.out.println("Problem is not solved.\n");
                    return false;
                }
            }
        }
        if(count==row*column) {
            System.out.println("Problem is solved!\n");
            return true;
        }
        else
            return false;
    }

    /**
     * method cell gives the index in given coordinates
     * @param index1 as row
     * @param index2 as column
     * @return int cell content at given indexes
     */
    @Override
    int cell(int index1, int index2) {
        if(index1<0 || index1>this.row || index2<0 || index2>this.column) {
            System.out.println("Invalid index values!!");
            System.exit(0);
        }
        return board[index1*this.column+index2];
    }

    /**
     * Abstract method equals,if two objects are the same,boards are the same
     * @param other AbstractBoard object
     * @return true if objects are same,else return false
     */
    @Override
    boolean equals(AbstractBoard other) {
        if ((other == null))
            return false;
        // If the object is compared with itself then return true
        if (other == this) {
            return true;
        }
        //Check if other is an instance of BoardArray2D or not
        if (!(other instanceof BoardArray1D)) {
            return false;
        }
        // typecast other so that we can compare data members
        BoardArray1D control = (BoardArray1D) other;
        if ((control.getColumn() != getColumn()) || (control.getRow() != getRow()))
            return false;
        if ((control.getColumn() == getColumn()) && (control.getRow() == getRow())) {
            for (int i = 0; i < getRow(); i++) {
                for (int j = 0; j < getColumn(); j++) {
                    if (board[i * getColumn() + j] != control.cell(i, j))
                        return false;
                }
            }
            return true;
        }
        return false;
    }

    @Override
    /**
     * getEmptyCellX method for using in global test function
     * @return int emptycell's x index
     */
    public int getEmptyCellX(){
        for(int i=0; i<getRow();i++){
            for(int j=0;j<getColumn();j++){
                if(board[((i*getColumn())+j)]==100)
                    return i;
            }
        }
        return -1;
    }
    /**
     * getEmptyCellY method for using in global test function
     * @return int emptycell's y index
     */
    @Override
    //Returns the coordinates of empty cell
    public int getEmptyCellY(){
        for(int i=0; i<getRow();i++){
            for(int j=0;j<getColumn();j++){
                if(board[((i*getColumn())+j)]==100)
                    return j;
            }
        }
        return -1;
    }
}

File: HW5/boardArray1D.cpp
/*File BoardArray1D.cpp
 *This is the implementation file for class BoardArray1D*/

#include "boardArray1D.h"

namespace EsraNameSpace{
	//No parameter constructor
	BoardArray1D::BoardArray1D(): row(9),column(9)
	{
		arr1D = new int[9*9];
		
		for(int i=0;i<row*column;i++){
				arr1D[i]=-1;
		}
		setNumOfBoard();
		setLastMove('S');
	}

	//Constructor takes row and column
	BoardArray1D::BoardArray1D(int nrow,int ncolumn): row(nrow),column(ncolumn)
	{
		arr1D = new int[row*column];

		for(int i=0;i<row*column;i++){
				arr1D[i]=-1;
		}
		setNumOfBoard();
		setLastMove('S');
	}

	//Constructor takes size
	BoardArray1D::BoardArray1D(int size): row(size),column(size)
	{
		arr1D = new int[size*size];
		
		for(int i=0;i<row*column;i++){
				arr1D[i]=-1;
		}
		setNumOfBoard();
		setLastMove('S');
	}

	//Destructor
	BoardArray1D::~BoardArray1D(){

		delete [] arr1D;
	}
	//Copy constructor
	BoardArray1D::BoardArray1D(const BoardArray1D& object){
		row=object.getRow();
		column=object.getColumn();
		numOfMovesPerformed=object.numberOfMoves();
		numOfBoardsCreated=object.numberOfBoards();
		lastMovePerf=object.lastMove();
		arr1D = new int[row*column];
		
		for(int i=0; i<row;i++){
			for(int j=0; j<column;j++){
				arr1D[i*column+j]=object.getIndex(i,j);
			}
		}

	}
	//Assignment operator
	BoardArray1D& BoardArray1D::operator =(const BoardArray1D& rightSide){
		//Error check for control objects are same size
		if (row= rightSide.getRow() && column==rightSide.getColumn()){
			for(int i=0; i<row;i++){
				for(int j=0; j<row;j++){
					arr1D[i*column+j]=rightSide.getIndex(i,j);
				}
			}
		}
		else
			cout<<"Objects are not same size!";
	}

	//Prints the board by sending it to cout
	void BoardArray1D::print(){

		for (int m = 0; m < row; m++){
	        for (int k = 0; k < column; k++) {
	          if(arr1D[m*column+k]!=100)
	            cout << arr1D[m*column+k] << " ";
	          else if(arr1D[m*column+k]==100)
	            cout<<"bb ";
	        }
	        cout<<endl;
	  }
	  cout<<endl;
	}

	//Reads file and determines new row and column informations
	void BoardArray1D::readFromFile(ifstream &input,string filename){
		
		input.open(filename.c_str());
		char readen[100];
	  	char *point;       //will used for strtok
	  	int i=0,j=0;

		if(input.fail()){
		    cerr<<"Error opening file!";
		    exit(1);
		}
		while(!input.eof()){

		    input.getline(readen,100);
		    point=strtok(readen," ");

		    while(point!=NULL){

		      point=strtok(NULL," ");
		      j++;
		    }
		    i++;
		}
		//determining row and column numbers according to file
		setRow(i);
		setColumn(j/i);
		input.close();
	}	

	//Reads file and assigns the values from file to board				
	void BoardArray1D::readAndAssign(ifstream &input,string filename){

		delete [] arr1D;

		arr1D = new int[getRow()*getColumn()];
		
		input.open(filename.c_str());
		std::string readen;
		int i=0,j=0;
		int number=0;
		if(input.fail()){
		cerr<<"Error opening file!";
		exit(1);
		}
		
		while(!input.eof()){
			for(i=0;i<row;++i){
			  for(j=0;j<column;++j){
			    input>>readen;
			    if(readen!="bb"){
			      //Readen is type of string so i convert it to integer
			      number=atoi(readen.c_str());
			     
			      arr1D[i*column+j]=number;
			    }
			    if(readen=="bb"){
			    	
			      arr1D[i*column+j]=100;
			    } 
			  }
			}
  		}

	}

	//Writes the board configuration to given file
	void BoardArray1D::writeToFile(ofstream &output,string &filename){
		
		output.open(filename.c_str()); 
		if(output.fail()){
			cerr<<"Error opening file!";
			exit(1);
		}
		if(output.is_open()){

			for(int i=0; i<row ;i++){
				for(int j=0; j<column; j++){
					if(arr1D[i*column+j]!=100 && arr1D[i*column+j]>9){
						output<<arr1D[i*column+j];
						if(j!=column-1)
							output<<" ";
					}

					else if(arr1D[i*column+j]<10){
						output<<"0"<<arr1D[i*column+j];
						if(j!=column-1)
							output<<" ";
					}
					else if(arr1D[i*column+j] == 100){
						output<<"bb";
						if(j!=column-1)
							output<<" ";
					}
				}
				if(i!=row-1)
					output<<"\n";
			}
		}
		else 
		cerr<<"Error opening file!\n";
		output.close();
	}	

	//Resets board to the solution
	void BoardArray1D::reset(){
		
		int num=1;
		for(int i=0;i<row;i++){
			for(int j=0; j<column;j++){
				if(i==row-1 && j==column-1 && arr1D[i*column+j]!=0){

					arr1D[i*column+j]=100;
				}
				else if(arr1D[i*column+j]!=0){

					arr1D[i*column+j]=num;
					num++;
				} 
			}
		}
	}

	//Sets the size and resets the board
	void BoardArray1D::setSize(int rowS,int columnS){
		
		setRow(rowS);
		setColumn(columnS);
		delete [] arr1D;

		arr1D = new int[getRow()*getColumn()];
		
		for(int i=0;i<row;i++){
			for(int j=0; j<column;j++){
				arr1D[i*column+j]=-1;
			}
		}
		reset();
	}

	//Moves the empty cell to the given direction
	int BoardArray1D::move(char direction){
		int emptCellX, emptCellY;
		int num;

		for(int i=0; i<row; i++){
			for(int j=0; j<column; j++){
				if(arr1D[i*column+j]==100){
					emptCellX=i;
					emptCellY=j;
				}
			}
		}
		if (direction=='U'){
			
			if(emptCellX==0 || arr1D[((emptCellX-1))*column+emptCellY]==0){
				return -1;
			}
			else if(emptCellX!=0 && arr1D[((emptCellX-1))*column+emptCellY]!=0){
				num=arr1D[((emptCellX-1))*column+emptCellY];
				arr1D[(emptCellX-1)*column+emptCellY]=100;
				arr1D[emptCellX*column+emptCellY]=num;
				setLastMove('U');
				setNumOfMove();
			}
			return 0;  
		}
		else if(direction== 'D'){

			if(emptCellX==row-1 || arr1D[(emptCellX+1)*column+emptCellY]==0){
				return -1; 
			}
			else if(emptCellX!=row-1 && arr1D[(emptCellX+1)*column+emptCellY]!=0){
				num=arr1D[(emptCellX+1)*column+emptCellY];
				arr1D[(emptCellX+1)*column+emptCellY]=100;
				arr1D[emptCellX*column+emptCellY]=num;
				setLastMove('D');
				setNumOfMove();
			} 
			return 0;
		}
		else if(direction=='R'){

			if(emptCellY==column-1 || arr1D[emptCellX*column+(emptCellY+1)]==0){
				return -1;
			}

			else if(emptCellY!=column-1 && arr1D[emptCellX*column+(emptCellY+1)]!=0){
				num=arr1D[emptCellX*column+(emptCellY+1)];
				arr1D[emptCellX*column+(emptCellY+1)]=100;
				arr1D[emptCellX*column+emptCellY]=num;
				setLastMove('R');
				setNumOfMove();
			} 
			return 0;
		}
		else if(direction=='L'){

			if(emptCellY==0 || arr1D[emptCellX*column+(emptCellY-1)]==0){ 
				return -1;
			}

			else if(emptCellY!=0 && arr1D[emptCellX*column+(emptCellY-1)]!=0){
				num=arr1D[emptCellX*column+(emptCellY-1)];
				arr1D[emptCellX*column+(emptCellY-1)]=100;
				arr1D[emptCellX*column+emptCellY]=num;
				setLastMove('L');
				setNumOfMove();
			} 
			return 0;
		}
	}

	//Shuffles the board
	int BoardArray1D::shuffle(){
		/*First reset the board to the solution*/
		reset();
		
		int randomMove;
		int control=0;
		int i=0;
		/*This loop written for shuffle the board making ramdom moves size*size.
		But if there is an impossible move, i call random func. again to make sure
		the board is shuffled properly */
		while( i<row*column){

		randomMove = (rand() % 4)+1 ;
		if(randomMove==1){
			control=move('U');
		} 
		if(randomMove==2){
			control=move('D');
		} 
		if(randomMove==3){
			control=move('L');
		} 
		if(randomMove==4){
			control=move('R');
		} 
		/*Control is the return value for U,D,R,L functions
		If control is -1 this means there is no room for that choice
		If control is 0 this means cell is moved and size*size number increases*/
		if(control==0)
			i++;
		}
		//cout<<"Shuffled version:\n";
		//print();
		return 0;
	}

	//Returns true if board is solved
	bool BoardArray1D::isSolved(){
		BoardArray1D controlObje(row,column);
		controlObje.reset();

		int count=0;
		for(int i=0; i<row; i++){
			for(int j=0; j<column; j++){
				if(arr1D[i*column+j]==controlObje.getIndex(i,j)){
					count++;
				}
				else if(arr1D[i*column+j]!=controlObje.getIndex(i,j)){
					cout<<"Problem is not solved.\n";
					return false;
				}
			}
		}
		if(count==row*column) {
			cout<<"Problem is solved!\n";
			return true;
		}
	}

	//Returns true if two boards are equal
	bool BoardArray1D::operator ==(const BoardArray1D& rightSide)  {
		int count=0;

		for(auto i=0; i<row;i++){
			for(auto j=0; j<column;j++){
				if(arr1D[(i*column)+j]==rightSide.getIndex(i,j)){
					count++;
				}
				else if(arr1D[(i*column)+j]!=rightSide.getIndex(i,j)){
					return false;
				}	
			}
		}
		if(count==row*column) 
			return true;	
	}

	/*() operator overloading, returns the number in given indexes if they're valid*/
	int BoardArray1D::operator ()(int index1,int index2){
		
		if((index1 >= 0 && index1 <= getRow()) && (index2>=0 && index2<=getColumn()))
			return arr1D[index1*column+index2];
		else 
			return -1;
	}

	//Returns the coordinates of empty cell
	void BoardArray1D::getEmptyCellX_Y(int &cellX,int &cellY){
		for(auto i=0; i<getRow();i++){
			for(auto j=0;j<getColumn();j++){
				if(arr1D[((i*getColumn())+j)]==100){
					cellX=i;
					cellY=j;
				}
			}
		}
	}
	
}//end of namespace